service/smq/client.go

```diff
@@ -0,0 +1,74 @@
+package smq
+// what a simple message queue could look like. We can use a queue to reduce the amount of writes to the DB
+// flush at certain intervals
+
+import (
+	"fmt"
+	"github.com/jackc/pgx/v4"
+	"github.com/mailgun/service/models"
+	"sync"
+	"time"
+)
+const(
+	DefaultDuration = time.Duration(15) * time.Second
+)
+
+var simpleMessageQueue *MessageQueue
+
+var (
+	smqInit sync.Once
+)
+
+
+type MessageQueue struct {
+	mu             sync.Mutex
+	CounterMap     map[string]*models.Event
+	sendTicker     *time.Ticker
+	db *pgx.Conn
+}
+
+func Default(db *pgx.Conn) *MessageQueue {
+	smqInit.Do(func() {
+		simpleMessageQueue = simpleMessageQueue.NewMessageQueue(db)
+	})
+	return simpleMessageQueue
+}
+
+
+func (m *MessageQueue) NewMessageQueue(db *pgx.Conn ) *MessageQueue {
+
+	ticker := time.NewTicker(DefaultDuration)
+	go func() {
```

Comment:
If the message queue was implemented how would you handle the situation where the queue contains `unflushed` items but the service is shutting down?

Reply:
A few ways I can think of.
One we could abstract the message queue to another pod, that way if the service shutdown that pod keeps going.
Two on service shutdown initiation we could have the service send a message to m.Flush() to flush the remaining data. However, we would need to update this methods return for any possible errors that may occur so the service can retry again if flush isn't sucessful